### DISCUSSION TITLE
Issue with custom provider against Discogs

### DISCUSSION THREAD

**Discussion by jmnunezizu** [nextauthjs/next-auth#5720] [Help]
### Question 💬

Hi,

I'm in the process of creating a new provider against Discogs (https://www.discogs.com/). Discogs authentication is done using OAuth 1.0a. Please see https://www.discogs.com/developers#page:authentication,header:authentication-oauth-flow.

The auth provider is successfully added to the auth page, and when I click the button, I am taken to the Discogs' authorise URL. E.g. `https://www.discogs.com/oauth/authorize?oauth_token=${token}&oauth_token_secret=${tokenSecret}&oauth_callback_confirmed=true`. The params have been redacted.

When I follow the authorise option, I am redirected back to my app under the URL `http://localhost:3000/api/auth/signin?error=Callback`.

In the log output, I see:

```
[next-auth][debug][GET_AUTHORIZATION_URL] {
  url: 'https://www.discogs.com/oauth/authorize?oauth_token={TOKEN}&oauth_token_secret={SECRET}&oauth_callback_confirmed=true',
  provider: {
    id: 'discogs',
    name: 'Discogs',
    type: 'oauth',
    version: '1.0',
    encoding: 'PLAINTEXT',
    authorization: { url: 'https://www.discogs.com/oauth/authorize', params: {} },
    accessTokenUrl: 'https://api.discogs.com/oauth/access_token',
    requestTokenUrl: 'https://api.discogs.com/oauth/request_token',
    profileUrl: 'https://api.discogs.com/oauth/identity',
    profile: [Function: profile],
    clientId: ${clientId},
    clientSecret: ${clientSecret},
    signinUrl: 'http://localhost:3000/api/auth/signin/discogs',
    callbackUrl: 'http://localhost:3000/api/auth/callback/discogs'
  }
}
```

```
[next-auth][error][OAUTH_V1_GET_ACCESS_TOKEN_ERROR]
https://next-auth.js.org/errors#oauth_v1_get_access_token_error undefined {
  statusCode: 401,
  data: 'Invalid signature. Expected signature base string: ydMdorWxQuiYqywLXGjgHQfctocZaDQG&ajRMHfrXLxSnKNfrwRAEoCuvHgeNGvlLDLIBpmTu'
}

[next-auth][error][OAUTH_CALLBACK_ERROR]
https://next-auth.js.org/errors#oauth_callback_error undefined {
  statusCode: 401,
  data: 'Invalid signature. Expected signature base string: ydMdorWxQuiYqywLXGjgHQfctocZaDQG&ajRMHfrXLxSnKNfrwRAEoCuvHgeNGvlLDLIBpmTu'
}
```

I could not find any information other than "if you're receiving this error, check that everything is set up correctly" in the docs. I also found https://github.com/nextauthjs/next-auth/issues/855, but that issue is supposed to be fixed now.

What could the problem be? I literally ran out of ideas.

Thanks for the help!

### How to reproduce ☕️

The code I have so far looks like this:

```
import { OAuthConfig, OAuthUserConfig } from 'next-auth/providers';

export interface DiscogsProfile extends Record<string, any> {
  id: number;
  username: string;
  resource_url: string;
  consumer_name: string;
}

export default function Discogs<P extends DiscogsProfile>(
  options: OAuthUserConfig<P>
): OAuthConfig<P> {
  return {
    id: 'discogs',
    name: 'Discogs',
    type: 'oauth',
    version: '1.0',
    encoding: 'PLAINTEXT',
    authorization: 'https://www.discogs.com/oauth/authorize',
    accessTokenUrl: 'https://api.discogs.com/oauth/access_token',
    requestTokenUrl: 'https://api.discogs.com/oauth/request_token',
    profileUrl: 'https://api.discogs/oauth/identity',
    profile(profile: P) {
      return {
        id: profile.id.toString(),
        name: profile.username,
        email: null,
        image: null,
      };
    },
    options,
  };
}
```

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by Roy-Ermers** [comment]
Did you find a solution?

**Reply by jmnunezizu** [reply]
Hi @Roy-Ermers – no, no solution yet. I know OAuth works fine in Discogs. I have successfully used it in the past with Passport.js using a provider I developed. See https://github.com/jmnunezizu/passport-discogs.

Unfortunately, I couldn't make it work with Next Auth.
Any help would be appreciated. 
Thanks.

**Reply by gcerame** [reply]
Still no solutions found? I'm having the exact same error and found your issue

**Reply by jmnunezizu** [reply]
@gcerame – nothing. Still awaiting a recommendation from the team. I could not pinpoint where the problem is.

**Reply by DavidOliverius** [reply]
@jmnunezizu Just having a look into this code as I'm trying to set up a Discogs provider on a personal project - you seem to be missing the .com for the `profileUrl`

I get a whole bunch of different errors now which I am working through but hopefully that helps in some way?

**Reply by jmnunezizu** [reply]
Hi @DavidOliverius – thanks for the input. It must've been some copy/paste error. I do have the `.com` in the `profileUrl` and get the same errors. I have updated the code in the question.
What errors are you seeing?

**Reply by DavidOliverius** [reply]
@jmnunezizu  I'm now getting a whole bunch of different errors with my Prisma set up, trying to get the user info into my DB. Using the T3 stack for the first time and not really familiar with much of it, so I'm banging my head against a wall trying to understand what is going on. I don't seem to be getting the `[OAUTH_V1_GET_ACCESS_TOKEN_ERROR]` anymore, but maybe that will resurface once I fix my Prisma errors.

**Reply by jmnunezizu** [reply]
@DavidOliverius do you get `GET_AUTHORIZATION_URL` printed in the console? next-auth prints the different steps. This is the only one that works for me. I am taken to the Discogs authorise page, accept it, and then I see the errors.

**Reply by DavidOliverius** [reply]
@jmnunezizu I have to admit that now I've fixed my Prisma DB errors, it's actually working for me. I can login through Discogs and everything seems to be working as expected. But I am using the T3 stack so not sure if that is doing something that perhaps your code isn't?

**Reply by jmnunezizu** [reply]
@DavidOliverius I am not really sure. I am using the custom provider I added to the question. I can use other providers (e.g., Facebook) without a problem. I can't imagine a part of my code that interacts with this workflow other than the custom provider.

**Reply by augustdolan** [reply]
@DavidOliverius could you post your provider config as well as what you did with Prisma to rectify the issue? To me it looks like this is an issue with the Discogs 1.0A version being out of spec and there being no intuitive way to intercept the headers sent out when `profileUrl` is used. That said, I would love to go through what you did to see if it somehow rectifies this.

Otherwise, anyone hold insight on how we might be able to to produce our own headers? Here's the lead I'm working off of: https://stackoverflow.com/questions/60437159/discogs-api-cant-get-anything-else-than-401-you-must-authenticate-to-access

My error is also technically different from OP's own. Mine reads: 

```
[next-auth][error][OAUTH_V1_GET_ACCESS_TOKEN_ERROR] 
https://next-auth.js.org/errors#oauth_v1_get_access_token_error undefined {
  statusCode: 401,
  data: '{"message": "You must authenticate to access this resource."}'
}
```

Note that the status code and error for mine and @jmnunezizu come from Discogs' response to the GET on `'https://api.discogs/oauth/identity'`

**Reply by bfourgeaud** [reply]
I am having the same problem as you guys.

@DavidOliverius could you share your provider config to test it and see what is not working on our side ?
@jmnunezizu Did you finally found any solution to this ?

**Comment by emulienfou** [comment]
Happy to share my working solution with you including email and image in profile:

```ts
import { OAuthConfig, OAuthUserConfig } from "next-auth/providers";

export interface DiscogsProfile extends Record<string, any> {
  id: number;
  username: string;
  resource_url: string;
  consumer_name: string;
}

const DiscogsProvider = <P extends DiscogsProfile>(options: OAuthUserConfig<P>): OAuthConfig<P> => ({
  id: "discogs",
  name: "Discogs",
  type: "oauth",
  version: "1.0",
  encoding: "PLAINTEXT",
  authorization: "https://www.discogs.com/oauth/authorize",
  accessTokenUrl: "https://api.discogs.com/oauth/access_token",
  requestTokenUrl: "https://api.discogs.com/oauth/request_token",
  profileUrl: "https://api.discogs.com/oauth/identity",
  async profile(profile, tokens) {
    const url = new URL(`https://api.discogs.com/users/${ profile.username }`);
    url.searchParams.set("oauth_consumer_key", options.clientId);
    url.searchParams.set("oauth_token", tokens.oauth_token as string);
    url.searchParams.set("oauth_token_secret", tokens.oauth_token_secret as string);

    const res = await fetch(url);
    const data = await res.json();

    return {
      id: profile.id.toString(),
      name: profile.username,
      email: data.email,
      image: data.avatar_url,
    };
  },
  style: {
    logo: "https://www.svgrepo.com/show/330306/discogs.svg",
    logoDark: "https://www.svgrepo.com/show/330306/discogs.svg",
    bg: "#fff",
    text: "#000000",
    bgDark: "#000000",
    textDark: "#fff",
  },
  options,
});

export default DiscogsProvider;

```
